### PR TITLE
Upgrade webpack to version 5

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no 
| New feature? |  yes
| BC breaks? | no 
| Deprecations? | no  
| Fixed tickets | parts of #6343 
| Related issues/PRs | ~~requires #6688~~, https://github.com/sulu/skeleton/pull/201 
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Upgrade webpack to version 5.

#### Why?

Support latest version of webpack. 

#### TODO

 - [x] Remove dependencies to `util`, `process` packages via `isemail` package.
 - [x] Remove dependency to `Buffer` package see own js
 - [x] Undo changes of https://github.com/sulu/sulu/pull/6690/commits/822516faa66cff14abd75c1fec9a3b79c0b12b06
 - [x] Fix Build with skeleton
 - [x] Replace Email Validator

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,18 +65,17 @@
         "identity-obj-proxy": "^3.0.0",
         "jest": "^26.4.2",
         "jest-canvas-mock": "^2.1.1",
-        "mini-css-extract-plugin": "^1.5.0",
+        "mini-css-extract-plugin": "^2.6.1",
         "mobx": "^4.0.0",
         "mobx-react": "^5.0.0",
         "moment-timezone": "^0.5.14",
-        "optimize-css-assets-webpack-plugin": "^6.0.1",
```
Comment (alexander-schranz) [inline package.json:78]: replaced by `css-minimizer-webpack-plugin`: https://github.com/NMFR/optimize-css-assets-webpack-plugin

File: package.json [hunk 2]
```diff
@@ -39,16 +39,16 @@
...
-        "@ckeditor/ckeditor5-dev-utils": "^26.1.0",
-        "@ckeditor/ckeditor5-theme-lark": "^31.0.0",
+        "@ckeditor/ckeditor5-dev-utils": "^30.3.2",
+        "@ckeditor/ckeditor5-theme-lark": "^34.2.0",
         "@wojtekmaj/enzyme-adapter-react-17": "^0.6.1",
         "autoprefixer": "^9.8.6",
         "babel-eslint": "^10.0.2",
         "babel-jest": "^26.3.0",
         "babel-loader": "^8.0.6",
-        "clean-webpack-plugin": "^4.0.0",
```
Comment (alexander-schranz) [inline package.json:53]: think we not longer need this with `clean: true`

File: package.json [hunk 3]
```diff
@@ -87,10 +86,9 @@
         "stylelint": "^13.2.1",
         "stylelint-config-standard": "^22.0.0",
         "url-search-params-polyfill": "^8.0.0",
-        "webpack": "^4.27.0",
-        "webpack-clean-obsolete-chunks": "^0.4.0",
```
Comment (alexander-schranz) [inline package.json:97]: The `clean: true` is the new way for this: https://github.com/GProst/webpack-clean-obsolete-chunks

File: webpack.config.js [hunk 4]
```diff
@@ -100,10 +102,6 @@ module.exports = (env, argv) => { // eslint-disable-line no-undef
                     use: [
                         {
                             loader: MiniCssExtractPlugin.loader,
-                            options: {
-                                // set path to public root from bundled css: https://github.com/sulu/sulu/pull/6225
-                                publicPath: path.relative(outputPath, '.'),
```
Comment (alexander-schranz) [inline webpack.config.js:105]: not longer required as all pathes in css/js are relative

**Comment by alexander-schranz** [comment]
While it did first look good to build with webpack 5 there are issues with the `isemail` package. There are currently 3 node packages which we require `util`, `process`, `Buffer`. This is not longer supported. The workarounds via webpack and browserify didn't work in combination with skeleton. So we will currently keep on webpack 4 until resolving this. Issues.

**Comment by niklasnatter** [comment]
I think I would be okay with replacing the `isemail` if this is a blocker in the future. Where did we use the `Buffer` class in out code? 🤔

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Url/Url.js [hunk 5]
```diff
@@ -4,8 +4,8 @@ import {observer} from 'mobx-react';
 import {action, computed, observable} from 'mobx';
 import classNames from 'classnames';
 import log from 'loglevel';
-import IsEmail from 'isemail';
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Url/Url.js:7]: Was never planned by the package to support browsers. They only support node based applications.

File: src/Sulu/Bundle/AdminBundle/Resources/js/utils/Email/validateEmail.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// @flow
+export default function(value: ?string): boolean {
+    if (!value) {
+        return false;
+    }
+
+    return /^(.+)@(.+)$/.test(value);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Resources/js/utils/Email/validateEmail.js]: This is very dump regex. Not sure how strict we should even implement things here. /cc @sulu/core-team 

We definitely should not be too strict in case of supporting also local and intranet domains `@localhost`, `@massive.local`, ...

File: package.json [hunk 7]
```diff
@@ -89,14 +88,12 @@
         "react-styleguidist": "^11.0.5",
         "react-test-renderer": "^17.0.0",
         "regenerator-runtime": "^0.13.3",
-        "style-loader": "^2.0.0",
```
Comment (alexander-schranz) [inline package.json:92]: I removed this as it seems not longer be used by us, think the css extract is working differently for us. And in the skeleton it was not even installed.

File: src/Sulu/Bundle/AdminBundle/Resources/js/utils/Email/validateEmail.js (hunk 6 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/utils/Email/validateEmail.js]: Looks like there is a official regex by whatwg:

 - https://html.spec.whatwg.org/#valid-e-mail-address
 
Its also used inside the safari webkit browser: https://github.com/WebKit/WebKit/blob/e205a844146bab749642d1b88974b706fe4d9e7e/Source/WebCore/html/EmailInputType.cpp#L38-L39

File: src/Sulu/Bundle/AdminBundle/Resources/js/utils/Ajv/formats/tests/idnEmailValidator.test.js [hunk 8]
```diff
@@ -5,10 +5,6 @@ test('Normal email address should pass validation', () => {
     expect(idnEmailValidator('hello@example.com')).toBe(true);
 });
 
-test('Email address with emojis should pass validation', () => {
-    expect(idnEmailValidator('🤌@😎')).toBe(true);
-});
-
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Resources/js/utils/Ajv/formats/tests/idnEmailValidator.test.js:11]: Test was introduced by @luca-rath in https://github.com/sulu/sulu/pull/5815 but even browsers yet does not support this. So I think we can safely remove it:

Bildschirmfoto 2022-11-30 um 01 21 48

File: UPGRADE.md [hunk 9]
```diff
@@ -2,6 +2,23 @@
 
 ## 2.6.0
 
+### Webpack 5 upgrade
+
+Sulu requires now new versions of the following JS packages:
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
Sulu now uses Webpack 5 to build the administration interface application. To enable this, the following JavaScript dependencies were updated/changed:
```

File: UPGRADE.md [hunk 10]
```diff
@@ -2,6 +2,23 @@
...
+- `webpack-manifest-plugin`: `^5.0.0`
+- `mini-css-extract-plugin`: `^2.7.1`
+- `optimize-css-assets-webpack-plugin` was removed replaced by `css-minimizer-webpack-plugin`: `^4.2.2`
+- `clean-webpack-plugin` was removed and replaced by `clean: true` webpack output option
+- `webpack-clean-obsolete-chunks` was removed and replaced by `clean: true` webpack output option
+- `is-email` was removed and replaced by `sulu-admin-bundle/utils/Email/validateEmail` method 
+
+If you created custom admin components you need to upgrade also
+that components. The core js will automatically be update
+via the [update build](https://docs.sulu.io/en/latest/upgrades/upgrade-2.x.html) command.
```
Comment (niklasnatter) [inline UPGRADE.md:20]: If you have integrated custom JavaScript components into the administration interface,
you might need to adjust your components to be compatible with the updated dependencies.
If you have not integrated custom JavaScript code, you project is adjusted automatically by the
[update build](https://docs.sulu.io/en/latest/upgrades/upgrade-2.x.html) command.

Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
If you have integrated custom JavaScript components into the administration interface,
you might need to adjust your components to be compatible with the updated dependencies.
If you have not integrated custom JavaScript code, you project is adjusted automatically by the
[update build](https://docs.sulu.io/en/latest/upgrades/upgrade-2.x.html) command.
```

File: package.json (hunk 7 above)
Comment (niklasnatter) [reply package.json:92]: maybe it was used by `react-styleguidist`? 🤔

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/blockCollection.scss [hunk 11]
```diff
@@ -42,7 +42,7 @@
 
 .selectMultipleButtonContainer {
     display: flex;
-    justify-content: end;
```
Comment (niklasnatter) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/blockCollection.scss:45]: was this change intentional?

Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/blockCollection.scss:45]: yes autoprefixer did complain that `flex-end` is supported better then `end` and has same effect and did output a warning message about this line.

File: webpack.config.js [hunk 12]
```diff
@@ -31,43 +30,47 @@ module.exports = (env, argv) => { // eslint-disable-line no-undef
...
-            new CleanObsoleteChunksPlugin(),
             new webpack.DefinePlugin({
                 SULU_ADMIN_BUILD_VERSION: JSON.stringify(suluVersion),
             }),
         ],
+        optimization: {
+            minimizer: [
+                // For webpack@5 you can use the `...` syntax to extend existing minimizers
+                //      see: https://webpack.js.org/plugins/css-minimizer-webpack-plugin/
+                `...`, // eslint-disable-line quotes
```
Comment (niklasnatter) [inline webpack.config.js]:
```suggestion
                '...', // extend existing minimizers: https://webpack.js.org/plugins/css-minimizer-webpack-plugin/
```

File: package.json (hunk 7 above)
Comment (niklasnatter) [reply package.json:92]: okay i think this was a leftover from a previous ckeditor configuration that was copied from their documentation: https://ckeditor.com/docs/ckeditor5/latest/installation/advanced/alternative-setups/integrating-from-source.html

File: webpack.config.js (hunk 4 above)
Comment (niklasnatter) [reply webpack.config.js:105]: this is nice 🥳 i think the reason we dont need it anymore is because we configure the `MiniCssExtractPlugin` with a relative `filename` option

File: package.json (hunk 7 above)
Comment (alexander-schranz) [reply package.json:92]: yes did see it in the ckeditor and I was thinking that I already once had that case. That I tried to use the style-loader but did then come to the conclution that it isn't compatible or require for our setup.

File: webpack.config.js [hunk 13]
```diff
@@ -31,43 +30,47 @@ module.exports = (env, argv) => { // eslint-disable-line no-undef
...
-            publicPath,
+            clean: true,
+            path: path.resolve(projectRootPath, publicDir, outputPath),
+            filename: '[name].[chunkhash].js',
         },
         stats: 'minimal',
         performance: {
             hints: false,
         },
         devtool: argv.mode === 'development' ? 'eval-source-map' : 'source-map',
```
Comment (niklasnatter) [inline webpack.config.js:53]: the build process works for me and the result looked good during my testing 🥳 
unfortunately, it looks like our `npm run watch` task does not work anymore. right now, i see the changes only after restarting the watch task

Comment (niklasnatter) [reply webpack.config.js:53]: the output of my console indicates that the change was detected, but no changes are written to the filesystem if i change the sulu source code

Comment (niklasnatter) [reply webpack.config.js:53]: okay i suspect that this is somehow caused by the location of the code of Sulu 😢 if i change the code of the `assets/admin/app.js` file, the watch task works as expected. maybe files are not reloaded if they are located within the `node_modules` folder?

Comment (niklasnatter) [reply webpack.config.js:53]: okay it looks like we are running into this issue: https://github.com/webpack/webpack/issues/11612
adding the following to the configuration fixes the behaviour of the watch task:
```
snapshot: {
    // detect changes in "node_modules/@sulu": https://github.com/webpack/webpack/issues/11612
    managedPaths: [], 
},
```

**Review by niklasnatter — APPROVED** [review]
🥳

**Comment by alexander-schranz** [comment]
🎉
